✨ Add MQT-internal MLIR dialect conversions

li-mingbao opened the pull request:
This PR introduces the dialect conversion between the MQTOpt dialect and the MQTDyn dialect of mqtcore in both directions.

## Checklist:

<!---

-->

- [x] The pull request only contains commits that are focused and relevant to this change.
- [x] I have added appropriate tests that cover the new/changed functionality.
- [x] I have updated the documentation to reflect these changes.
- [x] I have added entries to the changelog for any noteworthy additions, changes, fixes or removals.
- [x] I have added migration instructions to the upgrade guide (if needed).
- [x] The changes follow the project's style guidelines and introduce no new warnings.
- [x] The changes are fully tested and pass the CI checks.
- [x] I have reviewed my own code changes.

li-mingbao commented on mlir/lib/Conversion/MQT/MQTDynToMQTOpt/MQTDynToMQTOpt.cpp:
```diff
@@ -0,0 +1,647 @@
+                                                  oldQreg.getUsers().end());
+
+    // iterate over them and replace the operands with the new one
+    for (auto* user : llvm::reverse(qregUsers)) {
```
You are right the reverse does not matter here, especially as the users are not ordered anyway.
Regarding the extra vector, you cant modify any operations if you iterate over the users with value.getUsers() directly as this would break the iterator. Therefore, I have to copy the values beforehand in a vector.

li-mingbao commented on mlir/lib/Conversion/MQT/MQTOptToMQTDyn/MQTOptToMQTDyn.cpp:
```diff
@@ -0,0 +1,518 @@
+    if (oldQreg.use_begin() != oldQreg.use_end()) {
+      auto* oldQubitUser = oldQreg.use_begin().getUser();
+      oldQubitUser->replaceUsesOfWith(oldQreg, newQreg);
+    }
```
Regarding this, there are some cases with branches were there are multiple users of an mqtopt Value, so there isnt always just one user.

ystade commented on mlir/lib/Conversion/MQTDynToMQTOpt/MQTDynToMQTOpt.cpp:
```diff
@@ -0,0 +1,432 @@
+    auto optQreg = mqtoptOp.getQreg();
+
+    // put the pair of the dyn register and the latest opt register in the map
+    qRegMap.insert({dynQreg, optQreg});
```
Ah, I just saw that it is not an STL map. Perhaps the LLVM dense map does not have an `emplace` function. If that is the case, just ignore my comment.

ystade commented on mlir/lib/Conversion/MQTDynToMQTOpt/MQTDynToMQTOpt.cpp:
```diff
@@ -0,0 +1,432 @@
+        qregMap[dynQreg] = optInsertOp.getOutQreg();
+
+        // remove the qubits from the maps
+        qubitMap.erase(toErase);
```
May it be an idea to store a map from dyn qregs to a set all extracted qubits. Then, you could iterate over this set for the qreg you want to create the dealloc operation for and remove the respective qubit from the qubit map (as you are iterating over another container). AT the end, the qreg together with the set of extracted qubits can be erased from the new map in one go. This would also make the if branch in the loop superfluous.

ystade commented on mlir/lib/Conversion/MQTDynToMQTOpt/MQTDynToMQTOpt.cpp:
```diff
@@ -0,0 +1,432 @@
+
+    // append them to a single vector
+    std::vector<Value> allOptQubits(optInQubits.begin(), optInQubits.end());
+    allOptQubits.insert(allOptQubits.end(), optPosCtrlQubitsValues.begin(),
```
Just to be clear, here you will still need the insert to be able to insert a sequence of elements. That is not possible with emplace.

ystade commented on mlir/lib/Conversion/MQTDynToMQTOpt/MQTDynToMQTOpt.cpp:
```diff
@@ -0,0 +1,432 @@
+        optNegCtrlQubitsValues);
+
+    // iterate over all the dyn input qubits and update their latest opt qubit
+    Value dynQubit = nullptr;
+    for (size_t i = 0; i < dynAllQubits.size(); i++) {
+      dynQubit = dynAllQubits[i];
```
```suggestion
    for (size_t i = 0; i < dynAllQubits.size(); i++) {
      const auto& dynQubit = dynAllQubits[i];
```
In my opinion, it makes the code slightly clearer and I think the compiler can optimize this a bit better.

ystade commented on mlir/lib/Conversion/MQTOptToMQTDyn/MQTOptToMQTDyn.cpp:
```diff
@@ -0,0 +1,382 @@
+
+    Value optQubit = nullptr;
+    Value dynQubit = nullptr;
+    auto optResults = op->getResults();
```
Can the function `getAllOutQubits` be used here? I am asking because I find `getResults` very generic and one has to look into the definition of the operation what actually is the result. This applies to all other uses of `getResult` in both conversions as well.

DRovara commented:
> However, this conversion, i.e., changing the return type of the function `foo` is non-trivial. Especially, if the function `foo` already has other return values. @DRovara @flowerthrower Do you have any thoughts on that issue?

That's a good point. In my opinion, we will have to consider this situation for function calls and similar stuff for the MQTOpt dialect anyways. Consider for instance this situation in the [`quantum-sink` transformation](https://github.com/munich-quantum-toolkit/core/blob/main/mlir/test/Dialect/MQTOpt/Transforms/branch-opt.mlir#L31).
It isn't working directly with function calls, but it uses labels and jump operations for a similar idea.
Here, certain optimisations may also lead to changes in required parameters.

So, while it is not entirely the same situation, I believe discussing/finding a solution for these general cases may also help us decide what to do in the conversions.

ystade commented:
@li-mingbao Please finish this PR for the simple case for now. We will discuss internally how we want to solve the more complicated issues in general next week. For you, @li-mingbao , I think it makes more sense to focus on the simple case here and then set up the translation from and to QIR. I hope that sounds good to you. Let me know if you want to investigate something else. We are quite flexible here.

ystade commented:
@li-mingbao to be precise, you can ignore branching and function calls for now and just assume one main function with one code block containing quantum instructions.

DRovara commented on mlir/lib/Conversion/MQTDynToMQTOpt/MQTDynToMQTOpt.cpp:
```diff
@@ -0,0 +1,432 @@
+      : qReg(qReg), index(index), indexAttr(indexAttr) {}
+  QubitData() : qReg(nullptr), index(nullptr), indexAttr(nullptr) {}
+};
+// map each initial dyn qubit to its latest opt qubit
+llvm::DenseMap<mlir::Value, mlir::Value>& getQubitMap() {
+  static llvm::DenseMap<mlir::Value, mlir::Value> qubitMap;
+  return qubitMap;
+}
+// map each initial dyn register to its latest opt register
+llvm::DenseMap<mlir::Value, mlir::Value>& getQregMap() {
+  static llvm::DenseMap<mlir::Value, mlir::Value> qregMap;
+  return qregMap;
+}
+// map each initial dyn qubit to its metadata
+llvm::DenseMap<mlir::Value, QubitData>& getQubitDataMap() {
+  static llvm::DenseMap<mlir::Value, QubitData> qubitDataMap;
+  return qubitDataMap;
+}
```
Running the risk of making a misleading suggestion again:
Rather than using global/static variables which would likely break if we tried to run a conversion multiple times(?), wouldn't it make sense to store references to these maps as member variables, where the references are passed in the constructor?


The mqt-core roundtrip pass does something similar:
https://github.com/munich-quantum-toolkit/core/blob/main/mlir/lib/Dialect/MQTOpt/Transforms/ToQuantumComputationPattern.cpp#L44
Here, a reference to a `qc::QuantumComputation` is passed to the [constructor](https://github.com/munich-quantum-toolkit/core/blob/main/mlir/lib/Dialect/MQTOpt/Transforms/ToQuantumComputationPattern.cpp#L46) of the `RewritePattern`. The reference can then be modified inside the pattern and then used outside.

In this case, maybe one can pass references to all required maps in the constructors of the corresponding `ConversionPattern`s to have shared instances without using global variables.

ystade reviewed (commented):
This was again a very fruitful iteration. Thank you, @li-mingbao for your effort. Especially the passes are way cleaner now. I have put questions into the code where I am looking for an answer. Maybe you can provide some insights.

ystade commented on mlir/lib/Conversion/MQTOptToMQTDyn/MQTOptToMQTDyn.cpp:
```diff
@@ -0,0 +1,340 @@
+  matchAndRewrite(MQTGateOptOp op, typename MQTGateOptOp::Adaptor adaptor,
+                  ConversionPatternRewriter& rewriter) const override {
+    // get all the input qubits including the ctrl qubits
+    const auto dynInQubitsValues = adaptor.getInQubits();
+    const auto dynPosCtrlQubitsValues = adaptor.getPosCtrlInQubits();
+    const auto dynNegCtrlQubitsValues = adaptor.getNegCtrlInQubits();
```
Thanks for the clarification, that makes sense.

ystade reviewed (changes requested):
Hi @li-mingbao , we are approaching the goal here. Thanks for the continued efforts. This time I went through the code very thoroughly and commented on every detail I found. Still, I did not find too much anymore. If you addressed these changes, could you please ping or request a review from  @burgholzer as I think he wants to do a final round of review before this PR can be merged

ystade commented on mlir/test/Conversion/mqtdyn-to-mqtopt.mlir:
@DRovara would you mind to have a look over the testing setup for the conversions and check whether you have any comments?

ystade replied:
@li-mingbao to make sure you see this, I tag you here. Would you be so kind and realize these changes in the tests?

burgholzer replied:
No worries. I just did not make the connection 🙂
I'll resolve this for now.
One general comment for looking at the tests here remains anyway.

burgholzer replied:
Yeah, exactly that 👍🏻

ystade commented on mlir/lib/Conversion/MQTDynToMQTOpt/MQTDynToMQTOpt.cpp:
```diff
@@ -0,0 +1,467 @@
+    const auto optQreg = mqtoptOp.getQreg();
+
+    // put the pair of the dyn register and the latest opt register in the map
+    qregMap->try_emplace(dynQreg, optQreg);
```
Ok, that makes sense. Then just ignore all related comments and mark them as resolved.

DRovara reviewed (commented):
Thanks a lot @li-mingbao for the update on the tests! This actually looks super cool now, I really like how nice file-check seems for these sorts of tests!

I have made some minor comments but they are all really just nitpicks about the variable names - should be addressed pretty fast.

I can't really judge how well the tests cover the conversion code (my intuition says that for 2000 changed lines, this seems like a small-ish number of tests) but as long as we have no real codecov set up, that will always be hard to judge, and it's not like these are way too few test cases, so it should be fine.

burgholzer commented on mlir/lib/Conversion/MQTOptToMQTDyn/MQTOptToMQTDyn.cpp:
```diff
@@ -0,0 +1,340 @@
+
+#include "mlir/Conversion/MQTOptToMQTDyn/MQTOptToMQTDyn.h"
+
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/Dialect/Func/Transforms/FuncConversions.h"
```
```suggestion
#include <mlir/Dialect/Func/IR/FuncOps.h>
#include <mlir/Dialect/Func/Transforms/FuncConversions.h>
```

burgholzer commented on mlir/lib/Conversion/MQTOptToMQTDyn/MQTOptToMQTDyn.cpp:
```diff
@@ -0,0 +1,340 @@
+using namespace mlir;
+
+#define GEN_PASS_DEF_MQTOPTTOMQTDYN
+#include <mlir/Conversion/MQTOptToMQTDyn/MQTOptToMQTDyn.h.inc>
```
```suggestion
#include "mlir/Conversion/MQTOptToMQTDyn/MQTOptToMQTDyn.h.inc"
```

burgholzer commented on mlir/lib/Conversion/MQTOptToMQTDyn/MQTOptToMQTDyn.cpp:
```diff
@@ -0,0 +1,340 @@
+    // create result type
+    const auto& qregType = dyn::QubitRegisterType::get(rewriter.getContext());
+
+    // create new operation
+    auto mqtdynOp = rewriter.create<dyn::AllocOp>(
+        op.getLoc(), qregType, adaptor.getSize(), adaptor.getSizeAttrAttr());
+
+    const auto& optQreg = op.getQreg();
+    const auto& dynQreg = mqtdynOp.getQreg();
+
+    // update the operands of the opt register user
+    (*optQreg.getUsers().begin())->replaceUsesOfWith(optQreg, dynQreg);
+
+    // erase the old operation
+    rewriter.eraseOp(op);
```
This also seems to be a candidate for directly using `replace...`
That also holds for a couple more functions down below.

burgholzer commented on mlir/lib/Conversion/MQTOptToMQTDyn/MQTOptToMQTDyn.cpp:
```diff
@@ -0,0 +1,340 @@
+      // update the operands of the opt qubit user
+      (*optQubit.getUsers().begin())->replaceUsesOfWith(optQubit, dynQubit);
+
+      const std::vector<Operation*> bitUsers(oldBit.getUsers().begin(),
+                                             oldBit.getUsers().end());
+      // iterate over the users of the old bit and replace the old bit with the
+      // new bit
+      for (auto* user : bitUsers) {
+        user->replaceUsesOfWith(oldBit, newBit);
+      }
```
Iterator based for loop?

burgholzer commented on mlir/lib/Conversion/MQTOptToMQTDyn/MQTOptToMQTDyn.cpp:
```diff
@@ -0,0 +1,340 @@
+  LogicalResult
+  matchAndRewrite(MQTGateOptOp op, typename MQTGateOptOp::Adaptor adaptor,
+                  ConversionPatternRewriter& rewriter) const override {
+    // get all the input qubits including the ctrl qubits
+    const auto& dynInQubitsValues = adaptor.getInQubits();
+    const auto& dynPosCtrlQubitsValues = adaptor.getPosCtrlInQubits();
+    const auto& dynNegCtrlQubitsValues = adaptor.getNegCtrlInQubits();
+
+    // append them to a single vector
+    std::vector<Value> allDynInputQubits(dynInQubitsValues.begin(),
+                                         dynInQubitsValues.end());
+    allDynInputQubits.insert(allDynInputQubits.end(),
+                             dynPosCtrlQubitsValues.begin(),
+                             dynPosCtrlQubitsValues.end());
+    allDynInputQubits.insert(allDynInputQubits.end(),
+                             dynNegCtrlQubitsValues.begin(),
+                             dynNegCtrlQubitsValues.end());
```
Probably just me overlooking things, but the conversion in the other direction used a `getAllInQubits` method for this.
Shouldn't this be used here as well?

ystade replied:
Once again something I was wondering already: Here, you use the adaptor. For some reason the Adaptor of an operation does not implement all interface functions that are defined for an operation itself. Only the plain functions of the operation are available. However, the "but the conversion in the other direction used a getAllInQubits" made me curious. Is there really done exactly the same. Maybe, @li-mingbao  can shed some light on this question.

burgholzer replied:
Yeah. I still vaguely remember your comment from this PR here in my inbox. But the fact that this seems to work in the other conversion made me wonder.
Maybe one of the dialect definitions is missing something.

burgholzer commented on mlir/lib/Conversion/MQTOptToMQTDyn/MQTOptToMQTDyn.cpp:
```diff
@@ -0,0 +1,340 @@
+                             dynNegCtrlQubitsValues.end());
+
+    // get the static params and paramMask if they exist
+    DenseF64ArrayAttr staticParams = nullptr;
+    if (auto optionalParams = op.getStaticParams()) {
+      staticParams =
+          DenseF64ArrayAttr::get(rewriter.getContext(), optionalParams.value());
+    } else {
+      staticParams = DenseF64ArrayAttr{};
+    }
+    DenseBoolArrayAttr paramMask = nullptr;
+    if (auto optionalMask = op.getParamsMask()) {
+      paramMask =
+          DenseBoolArrayAttr::get(rewriter.getContext(), optionalMask.value());
+    } else {
+      paramMask = DenseBoolArrayAttr{};
+    }
```
Similar suggestion to the other file for simplifying this.

ystade commented on mlir/lib/Conversion/MQTDynToMQTOpt/CMakeLists.txt:
```diff
@@ -0,0 +1,16 @@
+
+get_property(CONVERSION_LIBS GLOBAL PROPERTY MLIR_CONVERSION_LIBS)
+list(APPEND CONVERSION_LIBS MQT::CoreIR)
+add_compile_options(-fexceptions)
```
Can/should we, however, add this to respective targets only, i.e., through `target_compile_options`or alike?

ystade commented on mlir/lib/Conversion/MQTDynToMQTOpt/MQTDynToMQTOpt.cpp:
```diff
@@ -0,0 +1,463 @@
+    // vectors
+    for (const auto& dynQubit : dynQubits) {
+      optQubits.emplace_back(qubitMapRef[dynQubit]);
+      qubitTypes.emplace_back(qubitType);
```
To make this more clear, I think, what @burgholzer wants to do instead is the following after the `optQubits` is created and fully populated. 
```c++
std::vector<Type> qubitTypes(optQubits.size(), qubitType);
```

ystade commented on mlir/lib/Conversion/MQTDynToMQTOpt/MQTDynToMQTOpt.cpp:
```diff
@@ -0,0 +1,463 @@
+      const auto& oldBit = oldBits[i];
+      const auto& newBit = newBits[i];
+
+      const std::vector<Operation*> bitUsers(oldBit.getUsers().begin(),
+                                             oldBit.getUsers().end());
+      // update the operand of the previous bit users
+      for (auto* user : bitUsers) {
+
+        user->replaceUsesOfWith(oldBit, newBit);
+      }
```
I knew that this come up again—I already asked the exact same question—that's actually the reason why I also go thorugh your feedback :D 
Short: No, you can't. The issue is that within the loop by calling `replaceUsesOfWith` you modify the iterator you want to iterate over. Let me propose to put a comment into the code that writes down the essence of your thread and my former thread about this issue.

burgholzer replied:
That makes sense. Had a very brief chat with copilot on that: https://copilot.microsoft.com/shares/1FNQF7PcQtUfmMUJDs18s
Of course, I don't know if it is hallucinating, but would it be worth trying its suggestion?
```c++
for (auto &use : llvm::make_early_inc_range(oldBit.getUses())) {
  use.getOwner()->replaceUsesOfWith(oldBit, newBit);
}
```

li-mingbao commented on mlir/lib/Conversion/MQTDynToMQTOpt/MQTDynToMQTOpt.cpp:
```diff
@@ -0,0 +1,463 @@
+    // extracted from the register
+    (*qregQubitsMap)[dynQreg].emplace_back(dynQubit);
+
+    // replace the old operation result with the new qubit otherwise there will
+    // be an unrealized conversion
+    rewriter.replaceOp(op, optQubit);
```
The `replaceOp` function replaces the result of the `op` with the new values given in the 2nd operand. You can normally use the new operation as the 2nd operand but the number of results have to match for that which is not the case here.

li-mingbao commented on mlir/lib/Conversion/MQTDynToMQTOpt/MQTDynToMQTOpt.cpp:
```diff
@@ -0,0 +1,463 @@
+      paramMask = DenseBoolArrayAttr{};
+    }
+
+    // create new operation
+    auto mqtoptOp = rewriter.create<MQTGateOptOp>(
+        op.getLoc(), ValueRange(optInQubits).getTypes(),
+        ValueRange(optPosCtrlQubitsValues).getTypes(),
+        ValueRange(optNegCtrlQubitsValues).getTypes(), staticParams, paramMask,
+        op.getParams(), optInQubits, optPosCtrlQubitsValues,
+        optNegCtrlQubitsValues);
+
+    const auto& optResults = mqtoptOp.getAllOutQubits();
+    // iterate over all the dyn input qubits and update their latest opt qubit
+    for (size_t i = 0; i < dynAllQubits.size(); i++) {
+      const auto& dynQubit = dynAllQubits[i];
+      qubitMapRef[dynQubit] = optResults[i];
+    }
+
+    // erase the old operation
+    rewriter.eraseOp(op);
```
I think the problem here is that `replaceOp` tries to replace the results of the old operation with the results of the new operation. Since the mqtdyn operation does not have any results, it cant really replace them, so I have to manually collect the results and use them later.

li-mingbao commented:
@burgholzer Thanks for the detailed feedback. I should have adressed most of the problems.
The CMakeLists files are probably still wrong as I am not really familiar with them and I would appreciate if you could help me with that.

Also I replaced most of the `eraseOp` and `replaceUsesWith` with `replaceOp` as you have suggested. This may look awkward in some instances as the operation does not really replace the operation itself but rather the uses of the results of the old operation with new values and erases the old operation afterwards. So for operations where the replacement operation has the same number of outputs this works pretty good but in other cases I have to use the specific values and not the operation itself as the operand.
In addition, this does not work for dyn gate operations as they do not have any results. 

And lastly, I added tests for the `barrierOp` and the `gphaseOp` but I have an issue with the syntax of the gphaseOp in the mqtopt dialect. 
From what I have seen the global phase gate takes a single angle and optionally controlled qubits as input.
So in the mqtopt dialect it should look like this. 
```
mqtopt.gphase(%cst)
%q1 = mqtopt.gphase(%cst) ctrl %q0 : ctrl !mqtopt.Qubit
```
 But neither of the expressions seems to work as I get the following error:
```
error: expected ':' followed by operation type
"mqtopt.gphase"(%cst) 
```
But even after adding the `:` the IR does not accept the operation.
In the mqtdyn dialect it seems to work but when I try to convert the operation to the mqtopt dialect I get the same statement as above and it does not work.

DRovara commented:
> And lastly, I added tests for the `barrierOp` and the `gphaseOp` but I have an issue with the syntax of the gphaseOp in the mqtopt dialect.

Thanks for bringing this up! We actually knew that the `gphaseOp` has some problems with parsing from the first batch of `dialect_feature.mlir` tests, but decided to postpone working on it. Turns out we ended up forgetting about it. 

@ystade, do you think you have time to resolve that problem? I guess since you're the one who set up the operations, it should probably be easiest for you. I can also look at it, but I'm not sure when I'll find the time to do that.

burgholzer reviewed (approved):
Many thanks @li-mingbao 🙏🏼 This looks good to me.
I'll leave the final say to @DRovara for this. If you are also happy, let's merge this.

burgholzer reviewed (commented):
I just noticed: This is definitely worth a changelog entry in the "Added" category. Let's not forget to add that before merging.

burgholzer commented:
> > Hey @DRovara, Each `extractQubit` operation in the mqtdyn dialect is converted to a mqtopt `extractQubit` operation and a mqtopt `insertQubit` operation. So if you extract a qubit twice then it will also be inserted twice. But the indices could still be wrong as I just take the indices from the mqtdyn operation. So the mqtopt operations could try to insert it to the same index twice.
> 
> 
> 
> Right, that's what I meant. But I think we decided on postponing that to a follow-up PR some time, so in my opinion, this is ready to merge now.
> 
> 
> 
> 

In that case, @li-mingbao can you please add a Changelog entry? Then, I'll merge.
